pkg/store/device/cache.go

```diff
@@ -142,6 +163,39 @@ func (c *networkChangeStoreCache) GetDevices() []*Info {
 	return devices
 }
 
+// Watch streams device cache updates to the caller
+// Unlike Watch on an Atomix store this watch has to take care that an event is
+// sent to each watch caller - hence the listener array
+// A replay option allows former entries to be replayed to the caller
+func (c *networkChangeStoreCache) Watch(ch chan<- stream.Event, replay bool) (stream.Context, error) {
+	c.mu.RLock()
+	_, ok := c.listeners[ch]
+	c.mu.RUnlock()
+	if ok {
+		return nil, fmt.Errorf("already listening to channel %v", ch)
+	}
+	if replay {
+		c.mu.RLock()
```

Comment:
There's a race condition with how this is ordered. Between replaying the existing devices and adding the new channel to `listeners`, events that are received will be missed by the new listener. We might want to just lock the entire replay when it's requested to ensure past events are serialized with future events.

Reply:
ok - will have a look

Reply:
OK - hopefully I have addressed it - I put a single lock around the replay and the adding of the listener - it should not be possible for the listen() loop to inject any new events while we are doing this because it will not be able to acquire the lock. I hope we cannot miss any - I presume the channel on the network change store will wait for the lock.